R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+    package$remote <- "CRAN"
+  }
+  package <- as.locked_package(package)
+  if (package$is_dependency_package && !package$name %in%
+    as.character(installed.packages(priority = "base")[,1])) {
+      package$latest_version <- package$latest_version_in_lockbox %||%
+        package$latest_version %||% get_latest_version(package)
+  }
+  package
+}
+
+get_latest_version <- function(package) {
+  if (package$remote == "CRAN") {
+    get_available_cran_version(package)
+  } else {
+    version_from_remote(package)
+  }
+}
+
+#' Combine two lists of dependencies via version comparisons.  Keep packages
+#' found in list1 on the left side of the entirety of list2, while moving
+#' the parent package to the space after it's rightmost dependency found in list2.
+combine_dependencies <- function(list1, list2, current_parent) {
+  names <- lapply(list(list1, list2), function(lst) vapply(lst, `[[`, character(1), "name"))
+
+  ## Certain packages are no longer on cran but incorporated into R Core
+  core_pkgs <- as.character(installed.packages(priority = "base")[,1])
+  list1 <- list1[!names[[1]] %in% core_pkgs]
+  names[[1]] <- names[[1]][!names[[1]] %in% core_pkgs]
+  list2 <- list2[!names[[2]] %in% core_pkgs]
+  names[[2]] <- names[[2]][!names[[2]] %in% core_pkgs]
+
+  if (length(list1) == 0) return(list2)
+  if (length(list2) == 0) return(list1)
+
+  names(list1) <- names[[1]]
+  names(list2) <- names[[2]]
+
+  ## Find the rightmost dependency of the current package and move our current
+  ## package to the immediate right of that spot.  This preserves previous
+  ## sorting order while ensuring that the current package will be installed
+  ## after all its dependencies
+  if (current_parent %in% names[[2]] && any(names[[2]] %in% names[[1]])) {
+    init_parent_slot <- which(names[[2]] == current_parent)
+    final_parent_slot <- max(which(names[[2]] %in% names[[1]]))
+    if (final_parent_slot > init_parent_slot) {
+      sel1 <- seq_along(names[[2]]) != init_parent_slot & seq_along(names[[2]]) <= final_parent_slot
+      sel2 <- seq_along(names[[2]]) > final_parent_slot
+      list2 <- c(list2[sel1], list2[init_parent_slot], list2[sel2])
+      names[[2]] <- c(names[[2]][sel1], names[[2]][init_parent_slot], names[[2]][sel2])
+    }
+  }
+
+  list2 <- swap_packages(names[[1]], names[[2]], list1, list2)
+
+  c(list1[!names[[1]] %in% names[[2]]], list2)
+}
+
+#' Swap packages by comparing version information. 1ist2 has already
+#' been ordered by dependency, so is imperative that we keep its order while
+#' potentially swapping in the corresponding packages in list1 that have later
+#' version requirements.
+swap_packages <- function(names1, names2, list1, list2) {
+  ## Swap packages when the package in list1 is more recent (> version) than
+  ## it's corresponding package in list2.  Keep the list 2 element if it is a
+  ## locked package, as well.  If their remotes are not identical and neither
```

Comment:
What if the remotes are identical and `list2`'s version is greater? Then don't swap...?